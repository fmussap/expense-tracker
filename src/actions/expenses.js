import database from '../firebase/firebase'
import * as actions from './actions-type'

export const createExpense = (expense) => {
  return ({
    type: actions.ADD_EXPENSE,
    expense
  })
}

export const addExpense = (expenseData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0
    } = expenseData
    const expense = { description, note, amount, createdAt }
    return database.ref(`users/${uid}/expenses`).push(expense)
      .then((ref) => {
        dispatch(createExpense({
          id: ref.key,
          ...expense
        }))
      })
      .catch((e) => {
        console.log('Error on add expense', e)
      })
  }
}

export const removeExpenseState = ({ id } = {}) => ({
  type: actions.REMOVE_EXPENSE,
  id
})

export const removeExpense = ({ id } = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid
    return database.ref(`users/${uid}/expenses/${id}`).remove()
      .then(() => {
        dispatch(removeExpenseState({ id }))
      })
      .catch((e) => {
        console.log('Error on remove expense', e)
      })
  }
}

export const editExpenseState = (id, updates) => ({
  type: actions.EDIT_EXPENSE,
  id,
  updates
})

export const editExpense = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid
    return database.ref(`users/${uid}/expenses/${id}`).set(updates)
      .then(() => {
        dispatch(editExpenseState(id, updates))
      })
      .catch((e) => {
        console.log('Error on edit expense', e)
      })
  }
}

export const setExpenses = (expenses) => ({
  type: actions.SET_EXPENSES,
  expenses
})

export const startSetExpenses = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid
    return database.ref(`users/${uid}/expenses`).once('value')
      .then((snapshot) => {
        const expenses = []
        snapshot.forEach((childSnapshot) => {
          expenses.push({
            id: childSnapshot.key,
            ...childSnapshot.val()
          })
        })
        dispatch(setExpenses(expenses))
      })
  }
}
