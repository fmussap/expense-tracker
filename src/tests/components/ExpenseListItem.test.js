import React from 'react'
import { shallow } from 'enzyme'

import ExpenseListItem from '../../components/ExpenseListItem'
import expenses from '../fixtures/expenses'

test('should render ExpenseList with expenses', () => {
  const wrapper = shallow(<ExpenseListItem {...expenses[0]} />)
  expect(wrapper).toMatchSnapshot()
})

test('should show a description of the expense', () => {
  const wrapper = shallow(<ExpenseListItem {...expenses[0]} />)
  expect(wrapper.find('h3').text()).toBe(expenses[0].description)
})

test('should show the amount and createdAt of the expense', () => {
  const wrapper = shallow(<ExpenseListItem {...expenses[0]} />)
  expect(wrapper.find('p').text()).toBe(`${expenses[0].amount} - ${expenses[0].createdAt}`)
})
