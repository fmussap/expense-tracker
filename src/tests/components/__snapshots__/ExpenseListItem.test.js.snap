// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseList with expenses 1`] = `
<div>
  <Link
    replace={false}
    to="/edit/1"
  >
    <h3>
      banana
    </h3>
  </Link>
  <p>
    $5.00
    -
    December 26th, 2017
  </p>
</div>
`;
